internal/internal_event_handlers.go

```diff
@@ -782,15 +788,51 @@ func (wc *workflowEnvironmentImpl) SideEffect(f func() (*commonpb.Payloads, erro
 	wc.logger.Debug("SideEffect Marker added", tagSideEffectID, sideEffectID)
 }
 
+func (wc *workflowEnvironmentImpl) lookupMutableSideEffect(id string) (*commonpb.Payloads, bool) {
```

Comment:
If I understand your algorithm correctly, this may be simpler (untested):

```go
func (wc *workflowEnvironmentImpl) lookupMutableSideEffect(id string) *commonpb.Payloads {
	// Fail if ID not found
	callCountPayloads := wc.mutableSideEffect[id]
	if len(callCountPayloads) == 0 {
		return nil
	}

	// Find the most recent call at/before the current call count
	var payloads *commonpb.Payloads
	payloadIndex := -1
	for callCount, maybePayloads := range callCountPayloads {
		if callCount <= currentCallCount && callCount > payloadIndex {
			payloads = maybePayloads
			payloadIndex = callCount
		}
	}
	return payloads
}
```

* No need for multiple iterations and sorting and all of that
* No need for a bool when nullness can represent it just fine
* I don't understand why you "garbage collect" old entries. Maybe I don't understand this map. Is call count per task or something else I am missing? If it's to help Go mem, no need (and it's actually not helping Go mem anyways tbh), and if it's to save the amount of iterations next time through, also no need.

Reply:
I'll test I think your code works and is simpler, thanks!

> Maybe I don't understand this map. Is call count per task or something else I am missing

It's call count per workflow execution. I have that documented here

>// Records the number of times a mutable side effect was called per ID over the
// life of the workflow. Used to help distinguish multiple calls to MutableSideEffect in the same
// WorkflowTask.
mutableSideEffectCallCounter map[string]int

But tbf it should be documented at `callCountPayloads` as well.

the other question is should it be per workflow execution or workflow task?

Reply:
Per entire workflow execution makes sense. I was trying to understand why you were "GC'ing" map elements instead of just leaving them there. But if it was just an attempt to help Go out, no worries (but don't need to do it).

Reply:
>But if it was just an attempt to help Go out, no worries (but don't need to do it)

Why do I not need to? If I don't the map will grow indefinitely and the memory will never be freed?

Reply:
> Why do I not need to?

Because the workflow execution struct is automatically collected with all fields when it leaves cache. I highly doubt storing every mutable side effect ever created for a specific workflow is harmful.

Also, many don't know this but deleting map elements does _not_ reclaim space in Go yet. And it's not like you're removing the nested map from its parent when it's empty anyways.

> If I don't the map will grow indefinitely and the memory will never be freed?

It won't grow indefinitely because it'll be bounded by the number of mutable side effects the user runs right? That also affects command slice size and other things harmlessly IMO.

Reply:
>Granted removing the reference to the payload pointer does allow it to be collected.

Yeah this is what I want to be collected because it could be larger